redbot/core/dev_commands.py

```diff
@@ -287,7 +270,10 @@ async def repl(self, ctx):
                     await ctx.send(self.get_syntax_error(e))
                     continue
 
-            variables["message"] = response
+            context = await ctx.bot.get_context(response)
+            env = self.get_environement(context)
+            env["__builtins__"] = __builtins__
+            env["_"] = previous
```

Comment:
I think this should only be happening when starting the REPL session (so before the whole `while` loop), rather than on each ran code piece. 

In case you want some reasoning:
a) this doesn't actually work as a session anymore as you're replacing any created variable on each code execution
b) this `context` will not actually be valid as it's not generated from a message using bot command
c) dynamically changing the variables in an active REPL session can result in unexpected behavior, as you've got two options - either you leave any variable that already got its value changed (and that is not gonna be easy to achieve and can be a bit confusing) or you override any variables in the environment with the values you get from `get_environement()` which will be annoying as it would take away the control of the values you have under different names in your REPL session

Reply:
TIL eval actually updates the env argument, that explains a lot of stuff